src/main/java/org/sagebionetworks/bridge/exporter/synapse/SynapseHelper.java

```diff
@@ -0,0 +1,545 @@
+     * @param tableId
+     *         Synapse table ID to upload the TSV to
+     * @param file
+     *         TSV file to apply to the table
+     * @return number of rows processed
+     * @throws BridgeExporterException
+     *         if there's a general error with Bridge EX
+     * @throws IOException
+     *         if there's an error uploading the file handle
+     * @throws SynapseException
+     *         if there's an error calling Synapse
+     */
+    public long uploadTsvFileToTable(String projectId, String tableId, File file) throws BridgeExporterException,
+            IOException, SynapseException {
+        // Upload TSV as a file handle.
+        FileHandle tableFileHandle = createFileHandleWithRetry(file, "text/tab-separated-values", projectId);
+        String fileHandleId = tableFileHandle.getId();
+
+        // start tsv import
+        CsvTableDescriptor tableDesc = new CsvTableDescriptor();
+        tableDesc.setIsFirstLineHeader(true);
+        tableDesc.setSeparator("\t");
+        String jobToken = uploadTsvStartWithRetry(tableId, fileHandleId, tableDesc);
+
+        // poll asyncGet until success or timeout
+        boolean success = false;
+        Long linesProcessed = null;
+        for (int loops = 0; loops < asyncTimeoutLoops; loops++) {
+            if (asyncIntervalMillis > 0) {
+                try {
+                    Thread.sleep(asyncIntervalMillis);
+                } catch (InterruptedException ex) {
+                    // noop
+                }
+            }
+
+            // poll
+            UploadToTableResult uploadResult = getUploadTsvStatus(jobToken, tableId);
+            if (uploadResult != null) {
+                linesProcessed = uploadResult.getRowsProcessed();
+                success = true;
+                break;
+            }
+
+            // Result not ready. Loop around again.
+        }
+
+        if (!success) {
+            throw new BridgeExporterException("Timed out uploading file handle " + fileHandleId);
+        }
+        if (linesProcessed == null) {
+            // Not sure if Synapse will ever do this, but code defensively, just in case.
+            throw new BridgeExporterException("Null rows processed");
+        }
+
+        return linesProcessed;
+    }
+
+    /**
+     * Appends the given row set to the given Synapse table. This is a retry wrapper.
+     *
+     * @param rowSet
+     *         row set to append
+     * @param tableId
+     *         Synapse table to appy it to
+     * @throws InterruptedException
+     *         if the async call is interrupted
+     * @throws SynapseException
+     *         if the Synapse call fails
+     */
+    @RetryOnFailure(attempts = 5, delay = 100, unit = TimeUnit.MILLISECONDS,
+            types = { InterruptedException.class, SynapseException.class }, randomize = false)
+    public void appendRowsToTableWithRetry(AppendableRowSet rowSet, String tableId) throws InterruptedException,
```

Comment:
Are there network timeouts talking to Synapse? They are wrapped in SynapseException I'm thinking? I would increase delay a little if so (250ms-500ms). Backing off would be better.